networkx/algorithms/summarization.py

```diff
@@ -210,3 +209,622 @@ def dedensify(G, threshold, prefix=None, copy=True):
             G.add_edge(compression_node, node)
         compressor_nodes.add(compression_node)
     return G, compressor_nodes
+
+
+def snap_edge_type_id(G, edge, edge_attributes):
+    """
+    Returns a identifier of the edge type
+
+    Creates an edge type ID based on the unique combination of values of its given
+    attributes
+
+    Parameters
+    ----------
+    G: graph
+        A Networkx graph
+    edge: tuple
+        A tuple of the source & target nodes in the graph
+    edge_attributes: iterable
+        The edge attributes that can be found in edges in the graph, and
+        are to be recognized in the summarization
+
+    Returns
+    -------
+    tuple: A representation of the specified edge type based on the attribute
+        values
+    """
+    if edge_attributes:
+        edge_attribute_values = []
+        edge_data = G.get_edge_data(*edge)
+        for attribute in edge_attributes:
+            attribute_value = edge_data.get(attribute)
+            edge_attribute_values.append(attribute_value)
+        edge_type_id = tuple(edge_attribute_values)
+    else:
+        edge_type_id = None
+    return edge_type_id
+
+
+def snap_calculate_bitmap(
+    G,
+    group_ids,
+    group_sets,
+    edge_attributes,
+    edge_index_lookup,
+    group_lookup=None,
+):
+    """
+    Produces the neighbor-groups bitmap and group-ground participation_counts data structures
+
+    Produces a bitmap indicating if a given node in the original graph has one
+    or more edges of a given edge type to nodes in another group.  Also produces a
+    group-group participation_counts indicating how many nodes in a group have edges with
+    node(s) with another group with a given edge type.  These structures are produced from the
+    original graph G and the group ids and group sets from the previous split operation in the
+    SNAP aggregation algorithm.
```

Comment:
@dschult Good idea.  I was able to remove the `bitmap` and `participation_counts` data structures in favor of the `neighbor_groups`.

I found a couple place to use dict-comprehensions, but there very well may be more.  Just let me know if I missed them, and I'll update them.

Reply:
I have not been able to find any alternatives to `group_lookup` without introducing more overhead for finding what group an node is a part of.  Same with the `edge_index_lookup`.  If you happen to know of way to mitigate those 

As you suggested, I was able to merge `group_ids` and `group_sets` into `groups` using dict` storage.

Reply:
I looked at this some -- mostly at `edge_index_lookup` to get rid of the index lookups.

I think that `snap_edge_type_id` is called too often. Can we start by constructing a dict keyed by edge to the `snap_edge_type`?  Then we replace function calls with dict lookups -- and we don't have to reconstruct the tuple each time we examine the edge.  Something like:  
```python
snap_edge_type = {edge: tuple(attrs.get(attr) for attr in edge_attributes) for edge, attrs in G.edges.items()}
```

Next, in `snap_calculate_neighbor_groups()`, where you construct a list of edge types for each neighbor group of a node, you could switch from a list of integers to a single `collections.Counter` object which makes counting easy while avoiding indexes. Counter is a dict that is set up for counting. You don't have to populate it with the possible values to be counted -- it just returns 0 if it hasn't counted that key before.  To use it, do something like:
```python
neighbor_groups[node] = {group_id: Counter() for group_id in groups}
```
Then where you use `edge_index_lookup[edge_type]`  you can use something like:
```python
neighbor_groups[node][neighbor_group_id][snap_edge_type[(node, neighbor)]] +=1
```
That saves the edge type construction and the edge index lookup (from an inner-inner-inner loop!). 
You use  `snap_edge_type[edge]` to give a tuple instead of computing the edge type and then looking up in `edge_index_lookup` to get an integer.

Also, in that function you can use `G.neighbors(node)` for both directed and undirected cases since `G.out_neighbors` is the same as `G.neighbors` for a `DiGraph`.

It looks like another inner loop would be simplified by using the dict `snap_edge_type` in the function that builds the graph, but I haven't dug into that one much yet.

Reply:
I just made these changes.  I also simplified the `snap_build_graph` as you suggested as well.